### PR TITLE
[Operator] Add diag op and related tests(#252)

### LINKED ISSUE #252: Code Contribution: 【Lv1】【Operator Development】diag
将以下算子贡献至https://github.com/FlagOpen/FlagGems/tree/master/src/flag_gems/ops 目录下

diag 

认领方式：
1. Issue中留言评论：「认领人名称+认领」
2. 在贡献者微信群内发送：认领的Issue号例如：「认领+[https://github.com/FlagOpen/FlagGems/issues/xxx」]

### PR DESCRIPTION
**Pull request by Gxiandy**
### PR Category

 Operator

### Type of Change

New Feature
### Description

Add diag op kernel and related tests
### Issue

Resolves #252 
### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### Performance

Operator: diag  Performance Test (dtype=torch.bfloat16, mode=cuda, level=comprehensive)
|Size    |     Torch Latency (ms)  |  Gems Latency (ms)  |       Gems Speedup    |     Size Detail|
|----------|------------------------------|-------------------------|------------------|-------|
|N/A               |    0.005120|            0.004096 |              1.250   |       ([torch.Size([64, 64])], {'diagonal': -4})|
|N/A                |   0.007168  |          0.006144   |            1.167     |     ([torch.Size([4096, 4096])], {'diagonal': 4})|
|N/A                 |  0.006144    |        0.006144     |          1.000       |   ([torch.Size([4])], {'diagonal': 4})|
|N/A                  | 0.012288      |      0.011264       |        1.091         | ([torch.Size([1024])], {'diagonal': 4})|

Operator: diag  Performance Test (dtype=torch.float16, mode=cuda, level=comprehensive)
|Size       |  Torch Latency (ms) |   Gems Latency (ms)  |       Gems Speedup   |      Size Detail|
|-----------|-------------------|-----------------------|------------------------|-------------|
|N/A      |             0.005120      |      0.004096      |         1.250   |       ([torch.Size([64, 64])], {'diagonal': 0})|
|N/A       |            0.007168       |     0.006144       |        1.167    |      ([torch.Size([4096, 4096])], {'diagonal': 3})|
|N/A       |            0.006144       |     0.006144       |        1.000     |     ([torch.Size([4])], {'diagonal': 4})|
|N/A        |           0.012288        |    0.021504       |        **0.571**      |    ([torch.Size([1024])], {'diagonal': 2})|

I don't know why the performance of the diag op implemented with Triton is worse than the PyTorch implementation when the dtype is float16 and diagonal is 3.

### CODE REVIEW

File: src/flag_gems/ops/diag.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def diag_1d_to_2d(x, diagonal=0):
+    N = x.shape[0]
+
+    if diagonal >= 0:
+        M = N + diagonal
+        output = torch.zeros((M, M), dtype=x.dtype, device=x.device)
+    else:
+        M = N - diagonal
```
Comment (StrongSpoon) [inline src/flag_gems/ops/diag.py:52]: M = N + abs(diagonal)

Comment (StrongSpoon) [inline src/flag_gems/ops/diag.py]: M = N + abs(diagonal)

File: src/flag_gems/ops/diag.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        output = torch.zeros((M, M), dtype=x.dtype, device=x.device)
+
+    stride = output.stride(0)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/diag.py]: since tensor output is newly initialized, we can assert its stride equal to M. while tensor x might not be contiguous. I think considering stride of tensor x to kernel makes sense.

File: src/flag_gems/ops/diag.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import triton.language as tl
+
+
+@triton.jit
+def diag_1d_to_2d_kernel(
+    data_ptr, output_ptr, N, M, stride, diagonal, BLOCK_SIZE: tl.constexpr
+):
+    idx = tl.program_id(0) * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+
+    if diagonal >= 0:
```
Comment (StrongSpoon) [inline src/flag_gems/ops/diag.py:12]: better hint diagonal as tl.constexpr. condition statement for non-constexpr is not performant in kernel.

File: src/flag_gems/ops/diag.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        row_idx = idx
+        col_idx = row_idx + diagonal
+    else:
+        col_idx = idx
+        row_idx = col_idx - diagonal
+
+    mask = (row_idx < M) & (col_idx >= 0) & (col_idx < M)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/diag.py]: why require col_idx >= 0?

File: src/flag_gems/ops/diag.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    else:
+        M = N - diagonal
+        output = torch.zeros((M, M), dtype=x.dtype, device=x.device)
+
+    stride = output.stride(0)
+    BLOCK_SIZE = 128
+
+    grid = lambda meta: (triton.cdiv(N, BLOCK_SIZE),)
+
+    diag_1d_to_2d_kernel[grid](x, output, N, M, stride, diagonal, BLOCK_SIZE=BLOCK_SIZE)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/diag.py]: add context
`with torch.cuda.device(x.device):`

File: src/flag_gems/ops/diag.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        diag_len = min(N, M - diagonal)
+    else:
+        diag_len = min(N + diagonal, M)
+    output = torch.empty(diag_len, dtype=x.dtype, device=x.device)
+    stride = x.stride(0)
+    BLOCK_SIZE = 128
+
+    grid = lambda meta: (triton.cdiv(N, BLOCK_SIZE),)
+
+    diag_2d_to_1d_kernel[grid](x, output, N, M, stride, diagonal, BLOCK_SIZE=BLOCK_SIZE)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/diag.py]: ditto

File: src/flag_gems/ops/diag.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def diag_2d_to_1d(x, diagonal=0):
+    N, M = x.shape
+    if diagonal >= 0:
+        diag_len = min(N, M - diagonal)
+    else:
+        diag_len = min(N + diagonal, M)
+    output = torch.empty(diag_len, dtype=x.dtype, device=x.device)
+    stride = x.stride(0)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/diag.py]: 2d tensor x could be non-contiguous. stride(0) and stride(1) are both needed.

File: tests/test_special_ops.py [hunk 8]
```diff
@@ -812,3 +812,19 @@ def test_accuracy_repeat_interleave_self_tensor(shape, dim, dtype):
...
     gems_assert_equal(res_out, ref_out)
+
+
+# Test diag op
+@pytest.mark.diag
+@pytest.mark.parametrize("shape", [(1024), (1024, 1024)])
+@pytest.mark.parametrize("diagonal", [-1, 0, 1])
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_accuracy_diag(shape, diagonal, dtype):
+    print(shape)
```
Comment (StrongSpoon) [inline tests/test_special_ops.py]: accuracy test doesn't need printing shape:)

File: tests/test_special_ops.py [hunk 9]
```diff
@@ -812,3 +812,19 @@ def test_accuracy_repeat_interleave_self_tensor(shape, dim, dtype):
     with flag_gems.use_gems():
         res_out = torch.repeat_interleave(inp, repeats, dim)
     gems_assert_equal(res_out, ref_out)
+
+
+# Test diag op
+@pytest.mark.diag
+@pytest.mark.parametrize("shape", [(1024), (1024, 1024)])
+@pytest.mark.parametrize("diagonal", [-1, 0, 1])
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
```
Comment (StrongSpoon) [inline tests/test_special_ops.py:823]: parameters need to cover more cases

File: benchmark/test_special_perf.py [hunk 10]
```diff
@@ -244,3 +244,23 @@ def upsample_nearest2d_input_fn(shape, dtype, device):
         dtypes=FLOAT_DTYPES,
     )
     bench.run()
+
+
+@pytest.mark.diag
+def test_perf_diag():
+    def diag_input_fn(shape, dtype, device):
+        if len(shape) == 1 or (len(shape) == 2 and shape[0] == shape[1]):
```
Comment (StrongSpoon) [inline benchmark/test_special_perf.py]: I think inputs should include non-square 2d tensors

**Review by StrongSpoon — COMMENTED** [review]
thanks for your contribution! I have some comments and please improve your code:)

File: benchmark/test_special_perf.py (hunk 10 above)
Comment (Gxiandy) [reply benchmark/test_special_perf.py]: torch.diag op does not support non-square 2d tensor

Comment (StrongSpoon) [reply benchmark/test_special_perf.py]: 截屏2024-11-12 12 43 48

actually it does. and I think your code supports non-square too.

Comment (Gxiandy) [reply benchmark/test_special_perf.py]: Ok ,I'll do it

File: src/flag_gems/ops/diag.py (hunk 4 above)
Comment (Gxiandy) [reply src/flag_gems/ops/diag.py:19]: not necessary

Comment (Gxiandy) [reply src/flag_gems/ops/diag.py]: not necessary

File: src/flag_gems/ops/diag.py (hunk 2 above)
Comment (Gxiandy) [reply src/flag_gems/ops/diag.py]: this stride is used to store data to output, I think it must be stride of tensor output

File: benchmark/test_special_perf.py [hunk 11]
```diff
@@ -244,3 +244,23 @@ def upsample_nearest2d_input_fn(shape, dtype, device):
...
+def test_perf_diag():
+    def diag_input_fn(shape, dtype, device):
+        if len(shape) == 1 or len(shape) == 2:
+            input = torch.randn(shape, device=device, dtype=dtype)
+            diagonal = random.randint(-4, 4)
+            yield input, {
+                "diagonal": diagonal,
+            },
+
+    bench = GenericBenchmark(
```
Comment (StrongSpoon) [inline benchmark/test_special_perf.py]: define bench as a GenericBenchmarkExcluse3D and test cases only include 1d and 2d shapes.

File: benchmark/test_special_perf.py [hunk 12]
```diff
@@ -244,3 +244,23 @@ def upsample_nearest2d_input_fn(shape, dtype, device):
...
+        input_fn=diag_input_fn,
+        op_name="diag",
+        torch_op=torch.diag,
+        dtypes=FLOAT_DTYPES,
```
Comment (StrongSpoon) [inline benchmark/test_special_perf.py]: diag supports float dtypes, int dtypes and bool.

File: src/flag_gems/ops/diag.py (hunk 2 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/diag.py]: kernel could reuse M as output stride. but for non-contiguous input tensor x, you need passing its stride into kernel.

File: src/flag_gems/ops/diag.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    else:
+        diag_len = min(N + diagonal, M)
+    if diag_len <= 0:
+        return torch.empty(0, dtype=x.dtype, device=x.device)
+    output = torch.empty(diag_len, dtype=x.dtype, device=x.device)
+    stride0 = x.stride(0)
+    stride1 = x.stride(1)
+    BLOCK_SIZE = 128
+
+    grid = lambda meta: (triton.cdiv(N, BLOCK_SIZE),)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/diag.py]: maybe triton.cdiv(diag_len, BLOCK_SIZE) is enough?

**Review by StrongSpoon — COMMENTED** [review]
good job! there left some issues to resolve;)

File: benchmark/test_special_perf.py (hunk 11 above)
Comment (Gxiandy) [reply benchmark/test_special_perf.py]: GenericBenchmarkExcluse3D is enough

File: benchmark/test_special_perf.py [hunk 14]
```diff
@@ -244,3 +245,23 @@ def upsample_nearest2d_input_fn(shape, dtype, device):
         dtypes=FLOAT_DTYPES,
     )
     bench.run()
+
+
+@pytest.mark.diag
+def test_perf_diag():
+    def diag_input_fn(shape, dtype, device):
+        if shape[0] < 10000:
```
Comment (StrongSpoon) [inline benchmark/test_special_perf.py]: why require shape[0] less than 10000? test shapes in GenericBenchmardExcluse3D all satisfy.

Comment (Gxiandy) [reply benchmark/test_special_perf.py]: sorry，it's my local requiement because of GPU memroy limits, forgot to remove it

**Review by StrongSpoon — APPROVED** [review]
lgtm
